notify/factory.go

```diff
@@ -27,57 +31,146 @@ import (
-	"email":                   wrap(email.New),
-	"googlechat":              wrap(googlechat.New),
-	"kafka":                   wrap(kafka.New),
-	"line":                    wrap(line.New),
-	"opsgenie":                wrap(opsgenie.New),
-	"pagerduty":               wrap(pagerduty.New),
-	"pushover":                wrap(pushover.New),
-	"sensugo":                 wrap(sensugo.New),
-	"slack":                   wrap(slack.New),
-	"teams":                   wrap(teams.New),
-	"telegram":                wrap(telegram.New),
-	"threema":                 wrap(threema.New),
-	"victorops":               wrap(victorops.New),
-	"webhook":                 wrap(webhook.New),
-	"wecom":                   wrap(wecom.New),
-	"webex":                   wrap(webex.New),
-}
-
-type NotificationChannel interface {
-	notify.Notifier
-	notify.ResolvedSender
-}
+// BuildReceiverIntegrations builds notifiers of the receiver and wraps each of them in Integration.
+func BuildReceiverIntegrations(
+	receiver GrafanaReceiverConfig,
+	tmpl *template.Template,
+	img images.ImageStore, // Used by some receivers to include as part of the source
+	newLogger logging.LoggerFactory,
+	newWebhookSender func(n receivers.Metadata) (receivers.WebhookSender, error),
+	newEmailSender func(n receivers.Metadata) (receivers.EmailSender, error),
+	orgID int64,
+	version string,
+) ([]*Integration, error) {
 
-func Factory(receiverType string) (func(receivers.FactoryConfig) (NotificationChannel, error), bool) {
-	receiverType = strings.ToLower(receiverType)
-	factory, exists := receiverFactories[receiverType]
-	return factory, exists
-}
+	type notificationChannel interface {
+		notify.Notifier
+		notify.ResolvedSender
+	}
 
-// wrap wraps the notifier's factory errors with receivers.ReceiverValidationError
-func wrap[T NotificationChannel](f func(fc receivers.FactoryConfig) (T, error)) func(receivers.FactoryConfig) (NotificationChannel, error) {
-	return func(fc receivers.FactoryConfig) (NotificationChannel, error) {
-		ch, err := f(fc)
-		if err != nil {
-			return nil, ReceiverValidationError{
-				Err: err,
-				// TODO it will be removed in the next PR
-				Cfg: &GrafanaReceiver{
-					UID:                   fc.Config.UID,
-					Name:                  fc.Config.Name,
-					Type:                  fc.Config.Type,
-					DisableResolveMessage: fc.Config.DisableResolveMessage,
-					Settings:              fc.Config.Settings,
-					SecureSettings:        nil,
-				},
+	var (
+		integrations      []*Integration
+		errors            types.MultiError
+		createIntegration = func(idx int, cfg receivers.Metadata, f func(logger logging.Logger) notificationChannel) {
+			logger := newLogger("ngalert.notifier."+cfg.Type, "notifierUID", cfg.UID)
+			n := f(logger)
+			i := NewIntegration(n, n, cfg.Type, idx)
+			integrations = append(integrations, i)
+		}
+		createIntegrationWithWebhook = func(idx int, cfg receivers.Metadata, f func(logger logging.Logger, w receivers.WebhookSender) notificationChannel) {
```

Comment:
Same here, a comment would help.
```suggestion
                // Helper function to create an integration for a notification channel that requires a webhook sender.
		createIntegrationWithWebhook = func(idx int, cfg receivers.Metadata, f func(logger logging.Logger, w receivers.WebhookSender) notificationChannel) {
```